### ISSUE TITLE
Updates doesn't decode virtual attribute

### ISSUE THREAD

**Issue by fengjh** [jinzhu/gorm#798]
I have a `User` model with a virtual attribute `Password` and I cannot use `Updates` to update the `Password`.

```go
var user User
if err := db.First(&user).Error; err != nil {
  panic(err)
}
originEncryptedPassword = user.EncryptedPassword
newAttrs = User{Name: "new Name", Password: "new Password"}

if err := db.Model(&user).Updates(newAttrs).Error; err != nil {
  panic(err)
}

if user.EncryptedPassword == originEncryptedPassword {
  log.Print("Update password failure")
}

// Update password failure
```

```go
type User struct {
	gorm.Model
	Name              string `sql:"not null"`
	Password          string `sql:"-"`
	EncryptedPassword string `sql:"not null" json:"-"`
}

// BeforeSave is a hook function provided by gorm package. It is used to:
// - update user password
func (user *User) BeforeSave(db *gorm.DB) (err error) {
	if user.Password != "" {
		if err = user.setEncryptedPassword(); err != nil {
			return
		}
	}

	return
}

func (user *User) setEncryptedPassword() error {
	pw, err := bcrypt.GenerateFromPassword([]byte(user.Password), 0)
	if err != nil {
		return err
	}
	user.EncryptedPassword = string(pw)
	user.Password = ""
	return nil
}
```

You can try to reproduce this problem with this demo: https://github.com/fengjh/gorm_cannot_updates_virtual_attributes

**Comment by jinzhu** [comment]
Fixed in v1.0_dev branch.

FYI, you must to use `scope.SetColumn("EncryptedPassword", string(pw))` to update columns in callbacks if want to save it into database.

**Comment by fengjh** [comment]
> Fixed in v1.0_dev branch.

Er.., but my test case didn't pass with the latest version of the `gorm` in `master` branch. 

```
➜  gorm_cannot_updates_virtual_attributes git:(master) go test -run TestUpdateUserPassword
--- FAIL: TestUpdateUserPassword (0.29s)
	main_test.go:54: unexpected error: invalid username or password
FAIL
exit status 1
FAIL	github.com/fengjh/gorm_cannot_updates_virtual_attributes	0.349s
```

- - - - - - - - - 

> FYI, you must to use scope.SetColumn("EncryptedPassword", string(pw)) to update columns in callbacks if want to save it into database.

Good to know. 👍 

I try to replace `AfterSave(db *gorm.DB)` with `AfterSave(scope *gorm.Scope)` and replace `user.EncryptedPassword = string(pw)` to `scope.SetColumn("EncryptedPassword", string(pw))`, then it works in this way. Thanks.
